microsoft/vscode
# feat(web): add support for language packs

## Description

This PR adds support for using language packs with VS Code Web via the CLI flag `--locale`.

## Known Limitations

There are a couple limitations with this implementation. Hoping that the Code Team can provide guidance on whether these limitations need to be addressed before merging and if so, how they recommend addressing them. 

### Implementation only works on a full build 

I am not sure why exactly, but this patch only works when you do a full build of VS Code and run that (meaning you can't use this in dev mode).

### Requires `commit` in `product.json`

I had to manually add commit to `product.json` in the build output because of [this logic](https://github.com/microsoft/vscode/blob/main/src/vs/server/node/remoteLanguagePacks.ts#L22).

### Only works with `--locale` CLI flag

Switching the display language only works using the `--locale` CLI flag. I believe this is because the `argv.json` is only modified in the client and saved in the IndexedDB. 

![image](https://user-images.githubusercontent.com/3806031/160485113-9ef519af-d38b-4fc0-b017-a2d48cece71b.png)

If we patched [this](https://github.com/microsoft/vscode/blob/e16d9ef81c4740b6f0a6afec86b69583067de0a8/src/vs/workbench/services/environment/browser/environmentService.ts#L61) to use the file system instead of `userRoamingDataHome`, it could potentially work normally (without passing a CLI flag). 

### Still shows "Install in Browser" which leads to error

Even if it's installed on the remote/server, you still see the "Install in Browser" button. When you click on it, you get an error toast notification. Error comes from [this block](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/extensionManagement/common/webExtensionManagementService.ts#L64-L66). Do language packs not have manifests?

![image](https://user-images.githubusercontent.com/3806031/160883552-db36a816-55a3-434d-a2a9-16d4ea89fce7.png)

![image](https://user-images.githubusercontent.com/3806031/160883480-b9d535b1-d47b-47ad-8192-0d27a1cd1d32.png)


## Demo

https://user-images.githubusercontent.com/3806031/160885289-7b5ceb6e-8c1d-4e11-8fdf-b3933f4bae62.mov

## How to Test
1. `git clone https://github.com/microsoft/vscode.git` 
2. `gh pr checkout 146012`
3. `yarn`
4. `yarn gulp "vscode-reh-web-darwin-x64-min"` (replace `darwin-x64` with your platform)
5. `cd .. &&  ./vscode-reh-web-darwin-x64/bin/code-server-oss` 
6. Open VS Code in browser
7. Download an extension pack (i.e. [Spanish](https://marketplace.visualstudio.com/items?itemName=MS-CEINTL.vscode-language-pack-es))
8. Install from VSIX file
9. Kill VS Code and run again:  `./vscode-reh-web-darwin-x64/bin/code-server-oss --locale es` 
10. See display in Spanish

Fixes #82595
Fixes https://github.com/coder/code-server/issues/4703

## Contributors

- @code-asher - wrote original patch
- @jsjoeio - extracted from code-server
